### PR TITLE
l2geth: fix `eth_getBlockRange`

### PR DESCRIPTION
**Pull request by tynes**
**Description**
Fix for `eth_getBlockRange` when an item is not found

### CODE REVIEW

File: l2geth/internal/ethapi/api.go [hunk 1]
```diff
@@ -780,7 +781,10 @@ func (s *PublicBlockChainAPI) GetBlockRange(ctx context.Context, startNumber rpc
 	// For each block in range, get block and append to array.
 	for number := startNumber; number <= endNumber; number++ {
 		block, err := s.GetBlockByNumber(ctx, number, fullTx)
-		if block == nil || err != nil {
+		if block == nil {
```
[the remark this answers is not in the document]
Comment (tynes) [reply l2geth/internal/ethapi/api.go:784]: ah good looks

**Review by Inphi — APPROVED** [review]

**Comment by mslipper** [comment]
Is there any way we could unit test this? And are we sure that the this new behavior doesn't break any implied contracts? Even though this is technically a bug, it's possible that the DTL or other downstream callers relied on the broken error handling.

File: l2geth/internal/ethapi/api.go [hunk 2]
```diff
@@ -780,9 +781,12 @@ func (s *PublicBlockChainAPI) GetBlockRange(ctx context.Context, startNumber rpc
 	// For each block in range, get block and append to array.
 	for number := startNumber; number <= endNumber; number++ {
 		block, err := s.GetBlockByNumber(ctx, number, fullTx)
-		if block == nil || err != nil {
+		if err != nil {
 			return nil, err
 		}
+		if block == nil {
+			return nil, ethereum.NotFound
```
Comment (tuxcanfly) [inline l2geth/internal/ethapi/api.go:788]: Why not use l2geth's `NotFound`?

https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/interfaces.go#L29

Comment (tuxcanfly) [reply l2geth/internal/ethapi/api.go:788]: It's easier for a client to error check using l2geth's `NotFound` rather than depending on geth.

Comment (smartcontracts) [reply l2geth/internal/ethapi/api.go:788]: Actually I think this should be a more aggressive custom error. Check above already confirms that the end of the block range is available, so if something in the middle is not available that suggests a more concerning problem. IMO this is a "should not happen" type of error.

Comment (tynes) [reply l2geth/internal/ethapi/api.go:788]: You are right that this should not use the L1 `NotFound`, my bad - `goimports` pulled that in on save and I should have double checked it. This is definitely a should not happen kind of error, so i can define a custom error type to return

**Comment by tynes** [comment]
> 

Unit tests wouldn't catch a downstream service hitting this error, what would happen is the dtl would index a `null` transaction

**Comment by tynes** [comment]
I've pushed a new commit that returns a custom error

**Review by smartcontracts — APPROVED** [review]

**Review by mslipper — APPROVED** [review]

**Pull request merged**